src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -89,5 +89,41 @@ public SourceGeneratedDocumentState WithUpdatedGeneratedContent(SourceText sourc
                 this.LanguageServices,
                 this.solutionServices);
         }
+
+        internal sealed class SourceGeneratedTextDocumentServiceProvider : IDocumentServiceProvider
+        {
+            public static readonly SourceGeneratedTextDocumentServiceProvider Instance = new();
+
+            private SourceGeneratedTextDocumentServiceProvider()
+            {
+            }
+
+            public TService? GetService<TService>()
+                where TService : class, IDocumentService
+            {
+                // right now, it doesn't implement much services but we expect it to implements all 
+                // document services in future so that we can remove all if branches in feature code
+                // but just delegate work to default document services.
+                if (SourceGeneratedDocumentOperationService.Instance is TService documentOperationService)
+                {
+                    return documentOperationService;
+                }
+
+                if (DocumentPropertiesService.Default is TService documentPropertiesService)
+                {
+                    return documentPropertiesService;
+                }
+
+                return null;
+            }
+
+            private class SourceGeneratedDocumentOperationService : IDocumentOperationService
+            {
+                public static readonly SourceGeneratedDocumentOperationService Instance = new();
+
+                public bool CanApplyChange => false;
+                public bool SupportDiagnostics => true;
```

Comment:
📝 The addition of `documentServiceProvider` for `SourceGeneratedDocumentState` is entirely to specify `SupportDiagnostics`, which otherwise defaults to false

Reply:
The previous comment was not correct. Here is the corrected comment:

📝 The addition of `documentServiceProvider` for `SourceGeneratedDocumentState` is entirely to specify `CanApplyChange`, which otherwise defaults to true